### ISSUE TITLE
Compilation fails with double precision variants

### ISSUE THREAD

**Issue by leroyvn** [mitsuba-renderer/mitsuba2#217]
## Summary

Compiling double-precision variants fails.

## System configuration

- Platform: macOS 10.15.5
- Compiler: Apple clang version 11.0.0 (clang-1100.0.33.17) (XCode 11.3)
- Python version: Python 3.7.7 :: Anaconda
- Mitsuba 2 version: 2.1.0 d66721d52d2483319f440c300f1e35198664abd6
- Compiled variants:
    * `scalar_rgb`
    * `scalar_rgb_double`

## Description

Compiling double-precision variants fails. The problematic files are:

- `projector.cpp`

```
[2/155] Building CXX object src/emitters/CMakeFiles/projector-obj.dir/projector.cpp.o
FAILED: src/emitters/CMakeFiles/projector-obj.dir/projector.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DLITTLE_ENDIAN -DMTS_ENABLE_PROFILER -D__OSX__ -I../include -I../ext/tinyformat -I../ext/enoki/include -I../ext/tbb/include -I../ext/pybind11/include -I/Users/vincent/miniconda3/envs/mitsuba2/include/python3.7m -flto=thin  -stdlib=libc++ -D_LIBCPP_VERSION -std=gnu++17 -fcolor-diagnostics -O3 -DNDEBUG -isysroot /Applications/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fPIC   -fno-stack-protector -fomit-frame-pointer -fno-math-errno -ffp-contract=fast -march=native -Wall -Wextra -Wno-unused-local-typedefs -fvisibility=hidden -MD -MT src/emitters/CMakeFiles/projector-obj.dir/projector.cpp.o -MF src/emitters/CMakeFiles/projector-obj.dir/projector.cpp.o.d -o src/emitters/CMakeFiles/projector-obj.dir/projector.cpp.o -c ../src/emitters/projector.cpp
../src/emitters/projector.cpp:99:30: error: no matching function for call to 'perspective_projection'
        m_camera_to_sample = perspective_projection(size, size, 0, m_x_fov,
                             ^~~~~~~~~~~~~~~~~~~~~~
../src/emitters/projector.cpp:221:19: note: in instantiation of member function 'mitsuba::Projector<double, mitsuba::Color<double, 3> >::Projector' requested here
MTS_EXPORT_PLUGIN(Projector, "Projection emitter")
                  ^
../include/mitsuba/render/sensor.h:196:1: note: candidate template ignored: deduced conflicting types for parameter 'Float' ('double' vs. 'float')
perspective_projection(const Vector<int, 2> &film_size,
^
1 error generated.
```
- `mesh.cpp`
```
[3/155] Building CXX object src/librender/CMakeFiles/mitsuba-render-obj.dir/mesh.cpp.o
FAILED: src/librender/CMakeFiles/mitsuba-render-obj.dir/mesh.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DLITTLE_ENDIAN -DMTS_BUILD_MODULE=MTS_MODULE_RENDER -DMTS_ENABLE_PROFILER -D__OSX__ -I../include -I../ext/tinyformat -I../ext/enoki/include -I../ext/tbb/include -I../ext/pybind11/include -I/Users/vincent/miniconda3/envs/mitsuba2/include/python3.7m -I../ext/rgb2spec -flto=thin  -stdlib=libc++ -D_LIBCPP_VERSION -std=gnu++17 -fcolor-diagnostics -O3 -DNDEBUG -isysroot /Applications/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fPIC   -fno-stack-protector -fomit-frame-pointer -fno-math-errno -ffp-contract=fast -march=native -Wall -Wextra -Wno-unused-local-typedefs -fvisibility=hidden -MD -MT src/librender/CMakeFiles/mitsuba-render-obj.dir/mesh.cpp.o -MF src/librender/CMakeFiles/mitsuba-render-obj.dir/mesh.cpp.o.d -o src/librender/CMakeFiles/mitsuba-render-obj.dir/mesh.cpp.o -c ../src/librender/mesh.cpp
../src/librender/mesh.cpp:328:18: error: cannot initialize a variable of type 'mitsuba::Mesh<double, mitsuba::Color<double, 3> >::ScalarFloat *' (aka 'double *') with an rvalue of type 'enoki::ArrayBase<float, enoki::DynamicArray<enoki::Packet<float, 8> > >::Value *' (aka 'float *')
    ScalarFloat *pos_out = mesh->m_vertex_positions_buf.data();
                 ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/librender/mesh.cpp:874:23: note: in instantiation of member function 'mitsuba::Mesh<double, mitsuba::Color<double, 3> >::build_parameterization' requested here
MTS_INSTANTIATE_CLASS(Mesh)
                      ^
1 error generated.
```

## Steps to reproduce

Compile Mitsuba with the aforementioned variants.

**Comment by leroyvn** [comment]
Any news about this issue? I'll be happy to help however possible.

**Comment by tatue64** [comment]
Unfortunately, the double variant seems not to be regularly tested. Just a hint for quick workarounds that worked for me:

In mesh.cpp 

in line 328 replace 
`ScalarFloat *pos_out = mesh->m_vertex_positions_buf.data();`

with

`ScalarFloat *pos_out = (ScalarFloat *) mesh->m_vertex_positions_buf.data();`

In projector.cpp

comment line 99 out

```
//        m_camera_to_sample = perspective_projection(size, size, 0, m_x_fov,
//                                                    1e-4f, 1e4f);
```

The third argument of perspective_projection, which gives the "offset", is invalid (0). I do not currently use the projector plugin, so I just comment this line out to compile the rest. 


Furthermore, I need to replace

```
--- a/include/mitsuba/render/optix/shapes.h
+++ b/include/mitsuba/render/optix/shapes.h
@@ -190,11 +190,12 @@ void prepare_ias(const OptixDeviceContext &context,
     unsigned int sbt_offset = base_sbt_offset;
 
     float T[12] = {
-        transf.matrix(0, 0), transf.matrix(0, 1), transf.matrix(0, 2), transf.matrix(0, 3),
-        transf.matrix(1, 0), transf.matrix(1, 1), transf.matrix(1, 2), transf.matrix(1, 3),
-        transf.matrix(2, 0), transf.matrix(2, 1), transf.matrix(2, 2), transf.matrix(2, 3)
+        float(transf.matrix(0, 0)), float(transf.matrix(0, 1)), float(transf.matrix(0, 2)), float(transf.matrix(0, 3)),
+        float(transf.matrix(1, 0)), float(transf.matrix(1, 1)), float(transf.matrix(1, 2)), float(transf.matrix(1, 3)),
+        float(transf.matrix(2, 0)), float(transf.matrix(2, 1)), float(transf.matrix(2, 2)), float(transf.matrix(2, 3))
     };

```

**Comment by leroyvn** [comment]
Thanks @tatue64, this fixes the problem for now!

**Comment by Speierers** [comment]
Thanks for reporting this. It should be fixed now: 1ee34ce7
